hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1633,12 +1815,36 @@ int SaraNcpClient::configureApn(const CellularNetworkConfig& conf) {
         }
     }
 
-    auto resp = parser_.sendCommand("AT+CGDCONT=%d,\"%s\",\"%s%s\"",
-            UBLOX_DEFAULT_CID, UBLOX_DEFAULT_PDP_TYPE,
-            (netConf_.hasUser() && netConf_.hasPassword()) ? "CHAP:" : "",
-            netConf_.hasApn() ? netConf_.apn() : "");
-    const int r = CHECK_PARSER(resp.readResult());
-    CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
+    char cgdcontVal[512] = {0};
+    char cgdcontFmt[20] = {0};
+    int rCgdcont = 0;
+    bool setApn = false;
+    snprintf(cgdcontFmt, sizeof(cgdcontFmt), "+CGDCONT: %d,%%511s", UBLOX_DEFAULT_CID);
```

Comment:
If we get more specific with it, I would argue that it's more likely to encounter a weird edge case.  The third field won't always match the APN exactly either, so we must use `strstr`.

Reply:
> The third field won't always match the APN exactly either

Can you please clarify what this means? `AT+CGDCONT?` output has fixed format `+CGDCONT: <cid>,<PDP_type>,
<APN>,...`. If what you mean is that the conifgured APN is modified with e.g. appended ICCID/IMSI (I think that's what I saw), I'd still prefer that we use `scanf` to just get the APN string and then perform partial match and account for whatever is appended by the modem instead of just blindly using `strstr` which will for example match `"apncustom"` with `"apn"`.

Please also add comments in the code on what the modem returns and why we need a partial match.

Reply:
That must be ok because on R510 the APN field gets modified from "super" to "super.mnc123.mcc732.gprs" for example.  It seems reasonable if it matches "super" or "10569.mcs" anywhere, it has been set previously.  I can add this as a comment if you are ok with that.

Reply:
I still don't understand why we cannot extract just the APN string with scanf and check whether it exactly matches what we expect or it matches `%s.mcc%*u.mnc%*u` instead of blindly doing `strstr` on the whole response line `+CGDCONT: 123,"IP","super.mcc123.mnc.456.gprsorwhateverelse"` which can easily lead to various incorrect matches?

Reply:
The main issue is the APN modification done by the R510 is not documented and thus might change.  I think we can be sure that the APN string will be contained somewhere within.  I suspect that it will always be first, but I do not know this to be certain, thus the current implementation.  I believe we can assume it will be first and always be followed with a \" or \.  since it's not modified by R510 until it registers and is not modified by R410.

If it doesn't find the APN it fails safe by setting the APN.  The harm would be finding the APN and the APN not being set correctly.  I think I've worked out a decent scanf() format based on the above and will submit this shortly.